### ISSUE TITLE
Wrong labels in BC, top of pre-release branch 

### ISSUE THREAD

**Issue by yzhwang** [gunrock/gunrock#196]
Here are the command lines that cause incorrect results for BC.
It appears for several option combinations when using 3 and 4 GPUs on all 14 datasets.

[bc.txt](https://github.com/gunrock/gunrock/files/507678/bc.txt)

**Comment by sgpyc** [comment]
The error is more likely to happen when frontier size is 1
```
/bin/test_bc_8.0_x86_64 smallworld --sw_scale=5 --sw_k=2 --device=0,1,2,3 --partition-seed=1475701595 --queue-sizing=5 --in-sizing=9
Generating Small World Graph ...
Small World Graph generated in 0.033 ms, k = 2, p = 0.000

Degree Histogram (32 vertices, 64 edges):
    Degree   0: 0 (0.00%)
    Degree 2^0: 0 (0.00%)
    Degree 2^1: 32 (100.00%)

Using 4 GPU(s): [ 0 1 2 3 ].
Partition begin. seed=1475701595
partition end. (11.165857 ms)
Using traversal-mode LB
1	   	   	 Thread started
3	   	   	 Thread started
0	   	   	 Thread started
2	   	   	 Thread started
__________________________
0	   	   	 Enact begins
1	   	   	 Enact begins
3	   	   	 Enact begins
2	   	   	 Enact begins
--------------------------
iteration 0 elapsed: 12.236118 ms, src = 0
Computing reference value ...
CPU BC finished in 0.002146 msec. Search depth: 17

Validity BC Value: 
INCORRECT: [1]: 3.000000 != 7.250000
result[...0.000000, 3.000000, 2.500000, 2.000000, 1.500000, 1.000000, ...]
reference[...0.000000, 7.250000, 6.750000, 6.250000, 5.750000, 5.250000, ...]
INCORRECT: [2]: 2.500000 != 6.750000
result[...0.000000, 3.000000, 2.500000, 2.000000, 1.500000, 1.000000, 0.500000, ...]
reference[...0.000000, 7.250000, 6.750000, 6.250000, 5.750000, 5.250000, 4.750000, ...]
INCORRECT: [3]: 2.000000 != 6.250000
result[...0.000000, 3.000000, 2.500000, 2.000000, 1.500000, 1.000000, 0.500000, 0.000000, ...]
reference[...0.000000, 7.250000, 6.750000, 6.250000, 5.750000, 5.250000, 4.750000, 4.250000, ...]
INCORRECT: [4]: 1.500000 != 5.750000
result[...0.000000, 3.000000, 2.500000, 2.000000, 1.500000, 1.000000, 0.500000, 0.000000, 3.500000, ...]
reference[...0.000000, 7.250000, 6.750000, 6.250000, 5.750000, 5.250000, 4.750000, 4.250000, 3.750000, ...]
INCORRECT: [5]: 1.000000 != 5.250000
result[...0.000000, 3.000000, 2.500000, 2.000000, 1.500000, 1.000000, 0.500000, 0.000000, 3.500000, 3.000000, ...]
reference[...0.000000, 7.250000, 6.750000, 6.250000, 5.750000, 5.250000, 4.750000, 4.250000, 3.750000, 3.250000, ...]
Number of errors occurred: 58

Validity Sigma: 
INCORRECT: [16]: 1.000000 != 2.000000
result[...1.000000, 1.000000, 1.000000, 1.000000, 1.000000, 1.000000, 1.000000, 1.000000, 1.000000, 1.000000, ...]
reference[...1.000000, 1.000000, 1.000000, 1.000000, 1.000000, 2.000000, 1.000000, 1.000000, 1.000000, 1.000000, ...]
Number of errors occurred: 2

Validity labels: 
INCORRECT: [7]: 2 != 7
result[...2, 3, 4, 5, 6, 2, 8, 9, 10, 11, ...]
reference[...2, 3, 4, 5, 6, 7, 8, 9, 10, 11, ...]
Number of errors occurred: 4

[0:1.000000,0.000000 1:1.000000,3.000000 2:1.000000,2.500000 3:1.000000,2.000000 4:1.000000,1.500000 5:1.000000,1.000000 6:1.000000,0.500000 7:1.000000,0.000000 8:1.000000,3.500000 9:1.000000,3.000000 10:1.000000,2.500000 11:1.000000,2.000000 12:1.000000,1.500000 13:1.000000,1.000000 14:1.000000,0.500000 15:1.000000,0.250000 16:1.000000,0.000000 17:1.000000,0.000000 18:1.000000,0.500000 19:1.000000,0.000000 20:1.000000,0.500000 21:1.000000,0.500000 22:1.000000,1.000000 23:1.000000,1.500000 24:1.000000,2.000000 25:1.000000,2.500000 26:1.000000,3.000000 27:1.000000,3.500000 28:1.000000,4.000000 29:1.000000,4.500000 30:1.000000,5.000000 31:1.000000,5.500000 ]

	Memory Usage(B)	 #keys0,0	 #keys0,1 #keys1,0	 #keys1,1	 #ins1,0	 #ins1,1 #keys2,0	 #keys2,1	 #ins2,0	 #ins2,1 #keys3,0	 #keys3,1	 #ins3,0	 #ins3,1 #keys4
GPU_0	 4194304	 42	 42	 17	 17	 27	 27	 12	 12	 18	 18	 12	 12	 18	 18	 77
GPU_1	 4194304	 42	 42	 27	 27	 45	 45	 22	 22	 36	 36	 27	 27	 45	 45	 112
GPU_2	 4194304	 42	 42	 7	 7	 9	 9	 17	 17	 27	 27	 32	 32	 54	 54	 92
GPU_3	 4194304	 42	 42	 17	 17	 27	 27	 22	 22	 36	 36	 32	 32	 54	 54	 107
	 queue_sizing =	 7.000000 	 7.000000	 in_sizing =	 9.000000

 [BC] finished.
 avg. elapsed: 12.2361 ms
 iterations: 1
 min. elapsed: 12.2361 ms
 max. elapsed: 12.2361 ms
 rate: 0.0105 MiEdges/s
 src: 0
 nodes_visited: 32
 edges_visited: 128
 nodes queued: 63
 edges queued: 64
 nodes redundance: 96.88%
 load time: 1238.1439 ms
 preprocess time: 66.9670 ms
 postprocess time: 29.6640 ms
 total time: 1349.2999 ms
```

**Comment by sgpyc** [comment]
The issue is caused by mis-alignment in data communication between two iteration loops. Mainly solved in a19245e by adding CPU sync points before and after the operations between iteration loops. Most of errors in runs listed in bc.txt are gone, left with the ones caused by out-of-memory.

The output is like this:
```
./bin/test_bc_8.0_x86_64 smallworld --sw_scale=5 --sw_k=2 --device=0,1,2,3 --partition-seed=1475701595 --queue-sizing=5 --in-sizing=9
Generating Small World Graph ...
Small World Graph generated in 0.034 ms, k = 2, p = 0.000

Degree Histogram (32 vertices, 64 edges):
    Degree   0: 0 (0.00%)
    Degree 2^0: 0 (0.00%)
    Degree 2^1: 32 (100.00%)

Using 4 GPU(s): [ 0 1 2 3 ].
Partition begin. seed=1475701595
partition end. (10.874033 ms)
Using traversal-mode LB
__________________________
--------------------------
iteration 0 elapsed: 11.442184 ms, src = 0
Computing reference value ...
CPU BC finished in 0.001907 msec. Search depth: 17

Validity BC Value: 
CORRECT
Validity Sigma: 
CORRECT
Validity labels: 
CORRECT
[0:1.000000,0.000000 1:1.000000,7.250000 2:1.000000,6.750000 3:1.000000,6.250000 4:1.000000,5.750000 5:1.000000,5.250000 6:1.000000,4.750000 7:1.000000,4.250000 8:1.000000,3.750000 9:1.000000,3.250000 10:1.000000,2.750000 11:1.000000,2.250000 12:1.000000,1.750000 13:1.000000,1.250000 14:1.000000,0.750000 15:1.000000,0.250000 16:2.000000,0.000000 17:1.000000,0.250000 18:1.000000,0.750000 19:1.000000,1.250000 20:1.000000,1.750000 21:1.000000,2.250000 22:1.000000,2.750000 23:1.000000,3.250000 24:1.000000,3.750000 25:1.000000,4.250000 26:1.000000,4.750000 27:1.000000,5.250000 28:1.000000,5.750000 29:1.000000,6.250000 30:1.000000,6.750000 31:1.000000,7.250000 ]

	Memory Usage(B)	 #keys0,0	 #keys0,1 #keys1,0	 #keys1,1	 #ins1,0	 #ins1,1 #keys2,0	 #keys2,1	 #ins2,0	 #ins2,1 #keys3,0	 #keys3,1	 #ins3,0	 #ins3,1 #keys4
GPU_0	 4194304	 42	 42	 17	 17	 27	 27	 12	 12	 18	 18 12	 12	 18	 18	 77
GPU_1	 4194304	 42	 42	 27	 27	 45	 45	 22	 22	 36	 36 27	 27	 45	 45	 112
GPU_2	 4194304	 42	 42	 7	 7	 9	 9	 17	 17	 27	 27 32	 32	 54	 54	 92
GPU_3	 4194304	 42	 42	 17	 17	 27	 27	 22	 22	 36	 36 32	 32	 54	 54	 107
	 queue_sizing =	 7.000000 	 7.000000	 in_sizing =	 9.000000

 [BC] finished.
 avg. elapsed: 11.4422 ms
 iterations: 1
 min. elapsed: 11.4422 ms
 max. elapsed: 11.4422 ms
 rate: 0.0112 MiEdges/s
 src: 0
 nodes_visited: 32
 edges_visited: 128
 nodes queued: 63
 edges queued: 64
 nodes redundance: 96.88%
 load time: 1091.8531 ms
 preprocess time: 74.6009 ms
 postprocess time: 29.6180 ms
 total time: 1209.6672 ms
```

**Comment by yzhwang** [comment]
Closed. For OOM issue, I've added it as a known issue in ae9079f2044462d6fa9229dd95774ef4e5be6f3a
